src/mono/mono/mini/jit-icalls.c

```diff
@@ -1743,6 +1743,11 @@ mini_init_method_rgctx (MonoMethodRuntimeGenericContext *mrgctx, MonoGSharedMeth
 													   mono_method_get_context (m), m->klass);
 		g_assert (data);
 
+		// we need a barrier before publishing data via mrgctx->infos [i] because the contents of data may not
+		//  have been published to all cores and another thread may read zeroes or partially initialized data
+		//  out of it, even though we have a barrier before publication of entries in mrgctx->entries below
+		mono_memory_barrier();
```

Comment:
Could an option be to use a local array in this loop and then do a memcpy into mrgctx->infos after the memory barrier we already fire below (where we copy entries)? That way we will only fire one memory barrier instead of info->num_entries.

Reply:
We will do a full memory barrier so it will have wider impact cross cpu caches. using local stack memory and memcpy won't prevent other cores from running at full speed, but since this is just on the init path and then never again, it might not be worth optimizing. You can keep as is.